### PR TITLE
admin add account : request times out #4920

### LINKED ISSUE #4920: admin add account : request times out
This happened 4 times in a row:

teammates.logic.core.Emails sendErrorReport: Crash report failed to send. Detailed error stack trace: 
javax.jdo.JDODataStoreException: Datastore Failure
	at org.datanucleus.jdo.NucleusJDOHelper.getJDOExceptionForNucleusException(NucleusJDOHelper.java:304)
	at org.datanucleus.jdo.JDOPersistenceManager.getObjectById(JDOPersistenceManager.java:1676)
	at org.datanucleus.jdo.JDOPersistenceManager.getObjectById(JDOPersistenceManager.java:1767)
	at org.datanucleus.store.appengine.jdo.DatastoreJDOPersistenceManager.getObjectById(DatastoreJDOPersistenceManager.java:73)
	at teammates.storage.api.AccountsDb.getAccountEntity(AccountsDb.java:212)
	at teammates.storage.api.AccountsDb.getAccount(AccountsDb.java:92)
	at teammates.storage.api.AccountsDb.getAccount(AccountsDb.java:104)
	at teammates.logic.core.AccountsLogic.isAccountAnInstructor(AccountsLogic.java:73)
	at teammates.logic.api.GateKeeper.isInstructor(GateKeeper.java:354)
	at teammates.logic.api.GateKeeper.getCurrentUser(GateKeeper.java:62)
	at teammates.logic.api.Logic.getCurrentUser(Logic.java:132)
	at teammates.ui.controller.Action.executeAndPostProcess(Action.java:343)
	at teammates.ui.controller.ControllerServlet.doPost(ControllerServlet.java:59)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:637)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:717)
	at org.mortbay.jetty.servlet.ServletHolder.handle(ServletHolder.java:511)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1166)
	at teammates.storage.datastore.DatastoreFilter.doFilter(DatastoreFilter.java:28)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.ParseBlobUploadFilter.doFilter(ParseBlobUploadFilter.java:125)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.runtime.jetty.SaveSessionFilter.doFilter(SaveSessionFilter.java:37)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.JdbcMySqlConnectionCleanupFilter.doFilter(JdbcMySqlConnectionCleanupFilter.java:60)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.TransactionCleanupFilter.doFilter(TransactionCleanupFilter.java:43)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at org.mortbay.jetty.servlet.ServletHandler.handle(ServletHandler.java:388)
	at org.mortbay.jetty.security.SecurityHandler.handle(SecurityHandler.java:216)
	at org.mortbay.jetty.servlet.SessionHandler.handle(SessionHandler.java:182)
	at org.mortbay.jetty.handler.ContextHandler.handle(ContextHandler.java:765)
	at org.mortbay.jetty.webapp.WebAppContext.handle(WebAppContext.java:418)
	at com.google.apphosting.runtime.jetty.AppVersionHandlerMap.handle(AppVersionHandlerMap.java:260)
	at org.mortbay.jetty.handler.HandlerWrapper.handle(HandlerWrapper.java:152)
	at org.mortbay.jetty.Server.handle(Server.java:326)
	at org.mortbay.jetty.HttpConnection.handleRequest(HttpConnection.java:542)
	at org.mortbay.jetty.HttpConnection$RequestHandler.headerComplete(HttpConnection.java:923)
	at com.google.apphosting.runtime.jetty.RpcRequestParser.parseAvailable(RpcRequestParser.java:78)
	at org.mortbay.jetty.HttpConnection.handle(HttpConnection.java:404)
	at com.google.apphosting.runtime.jetty.JettyServletEngineAdapter.serviceRequest(JettyServletEngineAdapter.java:148)
	at com.google.apphosting.runtime.JavaRuntime$RequestRunnable.run(JavaRuntime.java:468)
	at com.google.tracing.TraceContext$TraceContextRunnable.runInContext(TraceContext.java:437)
	at com.google.tracing.TraceContext$TraceContextRunnable$1.run(TraceContext.java:444)
	at com.google.tracing.CurrentContext.runInContext(CurrentContext.java:256)
	at com.google.tracing.TraceContext$AbstractTraceContextCallback.runInInheritedContextNoUnref(TraceContext.java:308)
	at com.google.tracing.TraceContext$AbstractTraceContextCallback.runInInheritedContext(TraceContext.java:300)
	at com.google.tracing.TraceContext$TraceContextRunnable.run(TraceContext.java:441)
	at com.google.apphosting.runtime.ThreadGroupPool$PoolEntry.run(ThreadGroupPool.java:235)
	at java.lang.Thread.run(Thread.java:745)
NestedThrowablesStackTrace:
com.google.appengine.api.datastore.DatastoreFailureException: The thread has been interrupted and the request has timed out
	at com.google.appengine.api.datastore.FutureHelper.getInternal(FutureHelper.java:81)
	at com.google.appengine.api.datastore.FutureHelper.quietGet(FutureHelper.java:63)
	at com.google.appengine.api.datastore.DatastoreServiceImpl.get(DatastoreServiceImpl.java:41)
	at org.datanucleus.store.appengine.RuntimeExceptionWrappingDatastoreService.get(RuntimeExceptionWrappingDatastoreService.java:56)
	at org.datanucleus.store.appengine.DatastorePersistenceHandler.get(DatastorePersistenceHandler.java:100)
	at org.datanucleus.store.appengine.DatastorePersistenceHandler.get(DatastorePersistenceHandler.java:112)
	at org.datanucleus.store.appengine.DatastorePersistenceHandler.fetchObject(DatastorePersistenceHandler.java:517)
	at org.datanucleus.state.JDOStateManagerImpl.validate(JDOStateManagerImpl.java:4263)
	at org.datanucleus.ObjectManagerImpl.findObject(ObjectManagerImpl.java:2444)
	at org.datanucleus.jdo.JDOPersistenceManager.getObjectById(JDOPersistenceManager.java:1671)
	at org.datanucleus.jdo.JDOPersistenceManager.getObjectById(JDOPersistenceManager.java:1767)
	at org.datanucleus.store.appengine.jdo.DatastoreJDOPersistenceManager.getObjectById(DatastoreJDOPersistenceManager.java:73)
	at teammates.storage.api.AccountsDb.getAccountEntity(AccountsDb.java:212)
	at teammates.storage.api.AccountsDb.getAccount(AccountsDb.java:92)
	at teammates.storage.api.AccountsDb.getAccount(AccountsDb.java:104)
	at teammates.logic.core.AccountsLogic.isAccountAnInstructor(AccountsLogic.java:73)
	at teammates.logic.api.GateKeeper.isInstructor(GateKeeper.java:354)
	at teammates.logic.api.GateKeeper.getCurrentUser(GateKeeper.java:62)
	at teammates.logic.api.Logic.getCurrentUser(Logic.java:132)
	at teammates.ui.controller.Action.executeAndPostProcess(Action.java:343)
	at teammates.ui.controller.ControllerServlet.doPost(ControllerServlet.java:59)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:637)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:717)
	at org.mortbay.jetty.servlet.ServletHolder.handle(ServletHolder.java:511)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1166)
	at teammates.storage.datastore.DatastoreFilter.doFilter(DatastoreFilter.java:28)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.ParseBlobUploadFilter.doFilter(ParseBlobUploadFilter.java:125)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.runtime.jetty.SaveSessionFilter.doFilter(SaveSessionFilter.java:37)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.JdbcMySqlConnectionCleanupFilter.doFilter(JdbcMySqlConnectionCleanupFilter.java:60)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.TransactionCleanupFilter.doFilter(TransactionCleanupFilter.java:43)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at org.mortbay.jetty.servlet.ServletHandler.handle(ServletHandler.java:388)
	at org.mortbay.jetty.security.SecurityHandler.handle(SecurityHandler.java:216)
	at org.mortbay.jetty.servlet.SessionHandler.handle(SessionHandler.java:182)
	at org.mortbay.jetty.handler.ContextHandler.handle(ContextHandler.java:765)
	at org.mortbay.jetty.webapp.WebAppContext.handle(WebAppContext.java:418)
	at com.google.apphosting.runtime.jetty.AppVersionHandlerMap.handle(AppVersionHandlerMap.java:260)
	at org.mortbay.jetty.handler.HandlerWrapper.handle(HandlerWrapper.java:152)
	at org.mortbay.jetty.Server.handle(Server.java:326)
	at org.mortbay.jetty.HttpConnection.handleRequest(HttpConnection.java:542)
	at org.mortbay.jetty.HttpConnection$RequestHandler.headerComplete(HttpConnection.java:923)
	at com.google.apphosting.runtime.jetty.RpcRequestParser.parseAvailable(RpcRequestParser.java:78)
	at org.mortbay.jetty.HttpConnection.handle(HttpConnection.java:404)
	at com.google.apphosting.runtime.jetty.JettyServletEngineAdapter.serviceRequest(JettyServletEngineAdapter.java:148)
	at com.google.apphosting.runtime.JavaRuntime$RequestRunnable.run(JavaRuntime.java:468)
	at com.google.tracing.TraceContext$TraceContextRunnable.runInContext(TraceContext.java:437)
	at com.google.tracing.TraceContext$TraceContextRunnable$1.run(TraceContext.java:444)
	at com.google.tracing.CurrentContext.runInContext(CurrentContext.java:256)
	at com.google.tracing.TraceContext$AbstractTraceContextCallback.runInInheritedContextNoUnref(TraceContext.java:308)
	at com.google.tracing.TraceContext$AbstractTraceContextCallback.runInInheritedContext(TraceContext.java:300)
	at com.google.tracing.TraceContext$TraceContextRunnable.run(TraceContext.java:441)
	at com.google.apphosting.runtime.ThreadGroupPool$PoolEntry.run(ThreadGroupPool.java:235)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.InterruptedException
	at com.google.common.util.concurrent.AbstractFuture.get(AbstractFuture.java:386)
	at com.google.common.util.concurrent.AbstractFuture$TrustedFuture.get(AbstractFuture.java:84)
	at com.google.appengine.tools.development.TimedFuture.get(TimedFuture.java:42)
	at com.google.common.util.concurrent.ForwardingFuture.get(ForwardingFuture.java:61)
	at com.google.appengine.tools.appstats.RecordingFuture.get(RecordingFuture.java:108)
	at com.google.appengine.tools.appstats.RecordingFuture$ExtendedFuture.get(RecordingFuture.java:49)
	at com.google.appengine.tools.appstats.RecordingFuture.get(RecordingFuture.java:108)
	at com.google.appengine.tools.appstats.RecordingFuture$ExtendedFuture.get(RecordingFuture.java:49)
	at com.google.appengine.api.utils.FutureWrapper.get(FutureWrapper.java:88)
	at com.google.appengine.api.datastore.AsyncDatastoreServiceImpl$5.getFutureWithOptionalTimeout(AsyncDatastoreServiceImpl.java:292)
	at com.google.appengine.api.datastore.AsyncDatastoreServiceImpl$5.aggregate(AsyncDatastoreServiceImpl.java:264)
	at com.google.appengine.api.datastore.AsyncDatastoreServiceImpl$5.get(AsyncDatastoreServiceImpl.java:235)
	at com.google.appengine.api.datastore.AsyncDatastoreServiceImpl$5.get(AsyncDatastoreServiceImpl.java:224)
	at com.google.appengine.api.utils.FutureWrapper.get(FutureWrapper.java:88)
	at com.google.appengine.api.utils.FutureWrapper.get(FutureWrapper.java:88)
	at com.google.appengine.api.utils.FutureWrapper.get(FutureWrapper.java:88)
	at com.google.appengine.api.datastore.FutureHelper.getInternal(FutureHelper.java:76)
	... 56 more

### PR DESCRIPTION
**Pull request by unyoungwax**
Fixes #4920

### CODE REVIEW

File: src/main/java/teammates/logic/backdoor/BackDoorLogic.java [hunk 1]
```diff
@@ -153,8 +154,17 @@ public String persistDataBundle(DataBundle dataBundle)
...
         for (FeedbackResponseAttributes response : responses.values()) {
-            updateRespondants(response.feedbackSessionName, response.courseId);
+            
+            String sessionString = response.feedbackSessionName + " % " + response.courseId;
+            
+            if (!sessionStrings.contains(sessionString)) {
+                updateRespondants(response.feedbackSessionName, response.courseId);
+            }
+            
+            sessionStrings.add(sessionString);
```
Comment (damithc) [inline src/main/java/teammates/logic/backdoor/BackDoorLogic.java]: Consider this:
```java
if (!sessionStrings.contains(sessionString)) {
    updateRespondants(response.feedbackSessionName, response.courseId);
    sessionStrings.add(sessionString);
 }
```
or this 
```java      
if (sessionStrings.contains(sessionString)) {
    continue;
}

updateRespondants(response.feedbackSessionName, response.courseId);
sessionStrings.add(sessionString);
```

The choice can be based on what you consider to be the 'normal flow'. Try to keep the normal flow un-nested and deal with 'abnormal' conditions using branches.

File: src/main/java/teammates/logic/backdoor/BackDoorLogic.java [hunk 2]
```diff
@@ -153,8 +154,17 @@ public String persistDataBundle(DataBundle dataBundle)
         }
         frDb.createFeedbackResponses(responses.values());
 
+        Set<String> sessionStrings = new HashSet<>();
+        
         for (FeedbackResponseAttributes response : responses.values()) {
-            updateRespondants(response.feedbackSessionName, response.courseId);
+            
+            String sessionString = response.feedbackSessionName + " % " + response.courseId;
```
Comment (kanghj) [inline src/main/java/teammates/logic/backdoor/BackDoorLogic.java]: @unyoungwax, it should be fine to not have the space between the %, to be consistent with <code>FeedbackSession.java</code>

**Comment by unyoungwax** [comment]
I have performed the requested changes.

File: src/main/java/teammates/logic/backdoor/BackDoorLogic.java [hunk 3]
```diff
@@ -153,8 +154,16 @@ public String persistDataBundle(DataBundle dataBundle)
         }
         frDb.createFeedbackResponses(responses.values());
 
+        Set<String> sessionStrings = new HashSet<>();
```
Comment (kanghj) [inline src/main/java/teammates/logic/backdoor/BackDoorLogic.java]: @unyoungwax, think of a better name than sessionStrings if you can

**Comment by kanghj** [comment]
@unyoungwax, let's try to fix this by today, passing AdminHomePageUiTest is too difficult without this.

File: src/main/java/teammates/logic/backdoor/BackDoorLogic.java (hunk 3 above)
Comment (unyoungwax) [inline src/main/java/teammates/logic/backdoor/BackDoorLogic.java]: How about `sessionIds`?

Comment (kanghj) [inline src/main/java/teammates/logic/backdoor/BackDoorLogic.java]: @unyoungwax, yes that should be fine

**Comment by bamboo3250** [comment]
It's much faster now :+1:

**Pull request merged**